include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -2934,6 +2934,347 @@ namespace Differentiation
+     *     // the tape.
+     *     ad_helper.stop_recording_operations(false); // write_tapes_to_file
+     *   }
+     *   else
+     *   {
+     *     // This is the "tape reuse" phase of the operations.
+     *     // Here we will leverage the already traced operations that reside
+     *     // on a tape, and simply re-evaluate the tape at a different point
+     *     // to get the function values and their derivatives.
+     *
+     *     // Load the existing tape to be reused
+     *     ad_helper.activate_recorded_tape(tape_index);
+     *
+     *     // Set the new values of the independent variables where the
+     *     // recorded dependent functions are to be evaluated (and
+     *     // differentiated around).
+     *     ad_helper.set_independent_variable(C, C_dofs);
+     *     ad_helper.set_independent_variable(H, H_dofs);
+     *   }
+     *
+     *   // Play the tape and record the output function value, its gradient and
+     *   // linearization. These are expensive to compute, so we'll do this once
+     *   // and extract the desired values from these intermediate outputs.
+     *   const double psi               = ad_helper.compute_value();
+     *   const Vector<double> Dpsi      = ad_helper.compute_gradient();
+     *   const FullMatrix<double> D2psi = ad_helper.compute_hessian();
+     *
+     *   // Extract the desired components of the gradient vector and Hessian
+     *   // matrix. In this example, we use them to compute the Piola-Kirchhoff
+     *   // stress tensor and its associated tangent, defined by thermodynamic
+     *   // arguments as S = 2*dpsi/dC and HH = 2*dS/dC...
+     *   const SymmetricTensor<2,dim> S =
+     *     2.0*ad_helper.extract_gradient_component(Dpsi,C_dofs);
+     *   const SymmetricTensor<4,dim> HH =
+     *     4.0*ad_helper.extract_hessian_component(D2psi,C_dofs,C_dofs);
+     *
+     *   // ... the magnetic induction and its associated tangent defined
+     *   // as B = -dpsi/dH and BB = dB/dH...
+     *   const Tensor<1,dim> B =
+     *     -ad_helper.extract_gradient_component(Dpsi,H_dofs);
+     *   const SymmetricTensor<2,dim> BB =
+     *     -symmetrize(ad_helper.extract_hessian_component(D2psi,H_dofs,H_dofs));
+     *
+     *   // ... and finally the magnetoelastic coupling tangent, defined
+     *   // as PP = -dS/dH = -d/dH(2*dpsi/dC). Here the order of the extractor
+     *   // arguments is especially important, as it dictates the order in which
+     *   // the directional derivatives are taken.
+     *   const Tensor<3,dim,double> PP =
+     *     -2.0*ad_helper.extract_hessian_component(D2psi,C_dofs,H_dofs)
+     * @endcode
+     *
+     * @warning ADOL-C does not support the standard threading models used by
+     * deal.II, so this class should @b not be embedded within a multithreaded
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
```

Comment:
For my own education, could one use a thread-local variable of these things, though? Or does ADOL-C keep global state? (I think I remember it being the latter.)

Reply:
I tried quite some time ago to implement a thread manager. ADOL-C has a global state and uses global variables, so unfortunately this simply doesn't work. Its really frustrating!